### PR TITLE
used convert_to_qmark_paramstyle function in odbc connector

### PR DESCRIPTION
**Pull request by raphaelvignes**
## Change Summary

Created a new Odbc Generic connector handling pyformat paramstyle in queries

## Checklist

* [x] Unit tests for the changes exist
* [x] Tests pass on CI and coverage remains at 100%
* [x] Documentation reflects the changes where applicable

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -25,6 +25,7 @@
     'mssql': ['pyodbc'],
     'mysql': ['PyMySQL>=0.8.0'],
     'odata': auth_deps + ['tctc_odata'],
+    'odbc': ['pyodbc', 'psycopg2'],
```
Comment (davinov) [inline setup.py:28]: Is `psycopg2` the postgres adaptor for odbc?
As it seems only used by tests, maybe it's a better idea to put it in `requirements-testing.txt`, so it does not get installed in production

File: tests/odbc/test_odbc.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+import pyodbc
+import pytest
+from pydantic import ValidationError
+
+from toucan_connectors.odbc.odbc_connector import OdbcConnector, OdbcDataSource
+
+
+def test_postgres_driver_installed():
+    """
+    This test aims to check that pgodbc is installed on the server
```
Comment (davinov) [inline tests/odbc/test_odbc.py:12]: I recommend keeping docstrings as short as possible, with direct active tone:
```suggestion
    Check that pgodbc is installed
```

File: tests/odbc/test_odbc.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+            database='postgres_db',
+            user='ubuntu',
+            password='ilovetoucan',
+        )
+        cur = conn.cursor()
+        cur.execute('SELECT 1;')
+        cur.close()
+        conn.close()
+
+    return service_container('postgres', check, psycopg2.Error)
```
Comment (davinov) [inline tests/odbc/test_odbc.py:32]: :+1:

File: tests/odbc/test_odbc.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+        name='test',
+        connection_string=(
+            'DRIVER={PostgreSQL Unicode};'
+            'DATABASE=postgres_db;'
+            'UID=ubuntu;'
+            'PWD=ilovetoucan;'
+            'SERVER=127.0.0.1;'
+            'PORT=' + str(postgres_server['port']) + ';'
+        ),
+    )
```
Comment (davinov) [inline tests/odbc/test_odbc.py:47]: Nice use of fixtures :+1:

File: tests/odbc/test_odbc.py [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+def test_odbc_get_df(mocker):
+    mockdbc = mocker.patch('pyodbc.connect')
+    mockdas = mocker.patch('pandas.read_sql')
+
+    odbc_connector = OdbcConnector(name='test', connection_string='blah')
+
+    ds = OdbcDataSource(
+        domain='test',
+        name='test',
+        query='SELECT Name, CountryCode, Population  from city LIMIT 2;',
```
Comment (davinov) [inline tests/odbc/test_odbc.py:70]:
```suggestion
        query='SELECT Name, CountryCode, Population from city LIMIT 2;',
```

File: tests/odbc/test_odbc.py [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+
+
+def test_raise_on_empty_query():
+    with pytest.raises(ValidationError):
+        OdbcDataSource(domain='test', name='test', database='postgres_db', query='')
+
+
+def test_odbc_get_df(mocker):
+    mockdbc = mocker.patch('pyodbc.connect')
+    mockdas = mocker.patch('pandas.read_sql')
```
Comment (davinov) [inline tests/odbc/test_odbc.py:63]: Don't hesitate to use variable identifiers that gives a good ideas of what's inside, and that are easily differentiable
```suggestion
    mock_pyodbc_connect = mocker.patch('pyodbc.connect')
    mock_pandas_read_sql = mocker.patch('pandas.read_sql')
```

File: tests/odbc/test_odbc.py [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+def test_retrieve_response(odbc_connector):
+    """ It should connect to the database and retrieve the response to the query """
+    ds = OdbcDataSource(query='select * from City;', domain='test', name='test')
+    res = odbc_connector.get_df(ds)
+    assert isinstance(res, pd.DataFrame)
+    assert res.shape[0] > 1
+
+
+def test_query_variability(mocker):
+    """ It should connect to the database and retrieve the response to the query """
```
Comment (davinov) [inline tests/odbc/test_odbc.py:88]: This test seems to be more about query parametrization
```suggestion
def test_query_parameters(mocker):
    """ It should send parametrized query with parameters in qmark style """
```

File: tests/test_common.py [hunk 8]
```diff
@@ -236,3 +237,33 @@ def test_connector_status():
...
+        (
+            'select * from test where id > %(id_nb)s and price > %(price)s;',
+            {'id_nb': 1},
+            'select * from test where id > ? and price > ?;',
+            [1, None],
+        ),
+    ],
+)
+def test_convert_pyformat_to_qmark(query, params, expected_query, expected_ordered_values):
+    """It should return query in qmark paramstyle and names of extracted params"""
```
Comment (davinov) [inline tests/test_common.py:266]:
```suggestion
    """It should return query in qmark paramstyle and values of extracted params"""
```

File: toucan_connectors/odbc/odbc_connector.py [hunk 9]
```diff
@@ -0,0 +1,44 @@
+import pandas as pd
+import pyodbc
+from pydantic import Field, constr
+
+from toucan_connectors.common import convert_to_qmark_paramstyle
+from toucan_connectors.toucan_connector import ToucanConnector, ToucanDataSource
+
+
+class OdbcDataSource(ToucanDataSource):
+    query: constr(min_length=1)
```
Comment (davinov) [inline toucan_connectors/odbc/odbc_connector.py:10]: :+1:

File: toucan_connectors/odbc/odbc_connector.py [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+
+
+class OdbcConnector(ToucanConnector):
+    """
+    Import data from ODBC apis
```
Comment (davinov) [inline toucan_connectors/odbc/odbc_connector.py:15]:
```suggestion
    Import data through ODBC API
```

File: toucan_connectors/odbc/odbc_connector.py [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+    """
+
+    data_source_model: OdbcDataSource
+
+    connection_string: str = Field(..., description='The connection string')
+    autocommit: bool = False
```
Comment (davinov) [inline toucan_connectors/odbc/odbc_connector.py:21]: As this connectors are only to read data, should we expose this parameter?
We could set it to `False` in the `connection_params` and not allow users to change it.

**Review by davinov — APPROVED** [review]
Thanks a lot! And congrats on your first connector :clap: 

The code is neat! Here is just a few style remarks, don't hesitate to question them if not clear or disagree.
None of them are blocking, so feel free to merge this PR once done

**Comment by raphaelvignes** [comment]
Thanks for all the comments, I've made all the relevant changes.

**Pull request merged**